### PR TITLE
stash: split into trait + impl

### PR DESCRIPTION
**Pull request by mjibson**
Make a Stash trait so it can be backed by other implementations.

### Motivation

   * This PR refactors existing code.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by mjibson** [comment]
Is this a good idea? I think this lets us experiment with dynamodb/etcd/fdb/postgres in a semi-reasonable way.

I'll rebase this on https://github.com/MaterializeInc/materialize/pull/11223 after it goes in.

**Comment by guswynn** [comment]
Ultimately all the impls are going to call `Codec::encode/decode`, so perhaps the trait should just offer `&[u8]` everywhere, and the `Stash` wrapper can do the `encode/decode`?

**Comment by mjibson** [comment]
I agree. Some of my local changes do exactly that. Will incorporate after 11223 merges.

File: src/stash/src/lib.rs
Comment (mjibson) [reply src/stash/src/lib.rs:52]: I have a followup PR that does this by forcing the trait impl to do the hard work: https://github.com/mjibson/materialize/commit/2af43f07d0e538764da0c19fe70650785f713e5d#diff-7de5a5c4a69f77e638b594788b2a3469e7e6c7bbd2cb1735f1027f228ffaca69R236

Comment (danhhz) [reply src/stash/src/lib.rs:52]: amazing 🙌 ! looks like `update_many` and `seal` are still around in that commit. we won't be able to implement them on the new persist api (which only has append style writes), so I think part of my question was if anyone had a sense for how hard it will be to migrate uses of them to `append`

Comment (mjibson) [reply src/stash/src/lib.rs:52]: Yeah we're in this exciting world where we have both the old and new api at once and it's not great. I'm going to work on this soon, but I think I can move the existing stash user (just timestamp bindings I think) to this new append api and then remove the old update/seal api which should unblock using storage as a stash impl.

Comment (mjibson) [reply src/stash/src/lib.rs:52]: coord tables have the same problem but are maybe getting fixed in https://github.com/MaterializeInc/materialize/pull/11572

Comment (danhhz) [reply src/stash/src/lib.rs:52]: awesome, i'm glad you're on it!

**Review by benesch — APPROVED** [review]

File: src/dataflow-types/src/client/controller/storage.rs
Comment (mjibson) [inline src/dataflow-types/src/client/controller/storage.rs:134]: Can a rust person tell me if this is ok or there's a better way? I tried to make `StorageControllerState` generic over `S` but then `StorageControllerState::new` complained that it was always producing a `Sqlite`, not an `S`. I then tried to put Box and dyn in various places unsuccessfully. I ended up with this because it compiled?

Comment (guswynn) [reply src/dataflow-types/src/client/controller/storage.rs:134]: I know we went with this, but for posterity, you can move the `new` method into an `impl StorageControllerState<Sqlite>` block, and leave everything else in the `impl< S: Stash> StorageControllerState<S>` block

**Pull request merged**
